### PR TITLE
Acpica pin

### PR DESCRIPTION
**Pull request by zetalog**
PinXxx resource terms support from Mika.
However, I still have questions needing Mika to reply.
So it's subject to change and do not merge it for now.

### CODE REVIEW

File: source/compiler/aslresources.y [hunk 1]
```diff
@@ -605,6 +606,25 @@ Memory32Term
...
+PinFunctionTerm
+    : PARSEOP_PIN_FUNCTION
+        PARSEOP_OPEN_PAREN          {$<n>$ = TrCreateLeafNode (PARSEOP_PIN_FUNCTION);}
+        OptionalShareType_First     /* 04: SharedType */
+        ',' PinConfigByte           /* 06: PinConfig */
+        ',' WordConstExpr           /* 08: FunctionNumber */
+        ',' StringData              /* 10: ResourceSource */
+        OptionalByteConstExpr       /* 11: ResourceSourceIndex */
+        OptionalResourceType        /* 12: ResourceType */
+        OptionalNameString          /* 13: DescriptorName */
```
Comment (zetalog) [inline source/compiler/aslresources.y]: DescriptorName is not in the spec.

File: source/compiler/aslrestype2s.c [hunk 2]
```diff
@@ -1339,3 +1340,189 @@ RsDoUartSerialBusDescriptor (
...
+
+    /*
+     * Calculate lengths for fields that have variable length:
+     * 1) Resource Source string
+     * 2) Vendor Data buffer
+     * 3) PIN (interrupt) list
+     */
+    ResSourceLength = RsGetStringDataLength (InitializerOp);
+    VendorLength = RsGetBufferDataLength (InitializerOp);
+    PinListLength = RsGetInterruptDataLength (InitializerOp, 8);
```
Comment (zetalog) [inline source/compiler/aslrestype2s.c]: That's why the previous commit is required.

File: source/compiler/aslrestype2s.c [hunk 3]
```diff
@@ -1339,3 +1340,189 @@ RsDoUartSerialBusDescriptor (
...
+            }
+            break;
+
+        case 5: /* Resource Usage (consumer/producer) */
+
+            /* Assumed to be consumer */
+
+            break;
+
+        case 6: /* Resource Tag (Descriptor Name) */
```
Comment (zetalog) [inline source/compiler/aslrestype2s.c]: Is this expected and the spec is then wrong?

File: source/compiler/asltokens.y [hunk 4]
```diff
@@ -389,6 +389,7 @@ NoEcho('
 %token <i> PARSEOP_PARITYTYPE_NONE
 %token <i> PARSEOP_PARITYTYPE_ODD
 %token <i> PARSEOP_PARITYTYPE_SPACE
+%token <i> PARSEOP_PIN_FUNCTION
```
Comment (zetalog) [inline source/compiler/asltokens.y]: Better naming should be PARSEOP_PINFUNCTION, like other resource descriptors.

File: tests/aslts/src/runtime/collections/functional/descriptor/pinfunction.asl [hunk 5]
```diff
@@ -0,0 +1,190 @@
+/*
+ * Some or all of this work - Copyright (c) 2006 - 2017, Intel Corp.
```
Comment (zetalog) [inline tests/aslts/src/runtime/collections/functional/descriptor/pinfunction.asl]: Corrected to 2017.

File: tests/aslts/src/runtime/collections/functional/descriptor/pinfunction.asl [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+/*
+ * Resource Descriptor macros
+ *
+ * PinFunction Resource Descriptor Macro
+ */
+
+Name (P45C, Package() {
+    ResourceTemplate () {
+        PinFunction(Exclusive, PullDefault, 0x1000, "\\_SB.GPO1", 0x0,
+            ResourceConsumer,, RawDataBuffer() {0xa, 0xb, 0xc, 0xd})
```
Comment (zetalog) [inline tests/aslts/src/runtime/collections/functional/descriptor/pinfunction.asl]: If there is no DescriptorName, this line should be:
+            ResourceConsumer, RawDataBuffer() {0xa, 0xb, 0xc, 0xd})

File: tests/aslts/src/runtime/collections/functional/descriptor/pinfunction.asl [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        PinFunction(Exclusive, PullDefault, 0x1234, "\\_SB.GPO1", 0x0,
+            ResourceConsumer, F0, RawDataBuffer() {0xa, 0xb, 0xc, 0xd})
+            {0x1, 0x2, 0x3}
+        PinFunction(Exclusive, PullDefault, 0x1234, "\\_SB.GPO1", 0x0,
+            ResourceConsumer, F1, RawDataBuffer() {0xa, 0xb, 0xc, 0xd})
+            {0x1, 0x2, 0x3}
+    }
+
+    m331(TS, 1, F0._SHR, 0x20, F1._SHR, 0x150, "_SHR")
+    m331(TS, 2, F0._PPI, 0x30, F1._PPI, 0x160, "_PPI")
```
Comment (zetalog) [inline tests/aslts/src/runtime/collections/functional/descriptor/pinfunction.asl]: No _FUN validation here?

File: source/compiler/aslresources.y [hunk 8]
```diff
@@ -606,6 +607,25 @@ Memory32Term
...
+PinConfigTerm
+    : PARSEOP_PIN_CONFIG
+        PARSEOP_OPEN_PAREN          {$<n>$ = TrCreateLeafNode (PARSEOP_PIN_CONFIG);}
+        OptionalShareType_First     /* 04: SharedType */
+        ',' ByteConstExpr           /* 06: PinConfigType */
+        ',' DWordConstExpr          /* 08: PinConfigValue */
+        ',' StringData              /* 10: ResourceSource */
+        OptionalByteConstExpr       /* 11: ResourceSourceIndex */
+        OptionalResourceType        /* 12: ResourceType */
+        OptionalNameString          /* 13: DescriptorName */
```
Comment (zetalog) [inline source/compiler/aslresources.y]: Similar issue, no DescriptorName in recent spec.

File: source/compiler/aslrestype2s.c [hunk 9]
```diff
@@ -1526,3 +1527,203 @@ RsDoPinFunctionDescriptor (
...
+            }
+            break;
+
+        case 5: /* Resource Usage (consumer/producer) */
+
+            RsSetFlagBits16 (&Descriptor->PinConfig.Flags, InitializerOp, 1, 1);
+
+            break;
+
+        case 6: /* Resource Tag (Descriptor Name) */
```
Comment (zetalog) [inline source/compiler/aslrestype2s.c]: Similar issue: no DescriptorName?

File: source/components/disassembler/dmresrc.c [hunk 10]
```diff
@@ -198,7 +198,8 @@ static ACPI_RESOURCE_HANDLER    AcpiGbl_DmResourceDispatch [] =
     AcpiDmExtendedDescriptor,       /* 0x0B, ACPI_RESOURCE_NAME_EXTENDED_ADDRESS_SPACE */
     AcpiDmGpioDescriptor,           /* 0x0C, ACPI_RESOURCE_NAME_GPIO */
     AcpiDmPinFunctionDescriptor,    /* 0x0D, ACPI_RESOURCE_NAME_PIN_FUNCTION */
-    AcpiDmSerialBusDescriptor       /* 0x0E, ACPI_RESOURCE_NAME_SERIAL_BUS */
+    AcpiDmSerialBusDescriptor,      /* 0x0E, ACPI_RESOURCE_NAME_SERIAL_BUS */
+    AcpiDmPinConfigDescriptor,      /* 0x0F, ACPI_RESOURCE_NAME_PIN_CONFIG */
```
Comment (zetalog) [inline source/components/disassembler/dmresrc.c]: Should be in alphanumeric order:
+    AcpiDmPinConfigDescriptor,      /* 0x0F, ACPI_RESOURCE_NAME_PIN_CONFIG */
+    AcpiDmSerialBusDescriptor,      /* 0x0E, ACPI_RESOURCE_NAME_SERIAL_BUS */

File: tests/aslts/src/runtime/collections/functional/descriptor/pinconfig.asl [hunk 11]
```diff
@@ -0,0 +1,537 @@
+/*
+ * Some or all of this work - Copyright (c) 2006 - 2017, Intel Corp.
```
Comment (zetalog) [inline tests/aslts/src/runtime/collections/functional/descriptor/pinconfig.asl]: Corrected to 2017.

File: source/common/ahpredef.c [hunk 12]
```diff
@@ -444,6 +444,7 @@ const AH_PREDEFINED_NAME    AslPredefinedInfo[] =
     AH_PREDEF ("_UPC",    "USB Port Capabilities", "Returns a list of USB port capabilities"),
     AH_PREDEF ("_UPD",    "User Presence Detect", "Returns user detection information"),
     AH_PREDEF ("_UPP",    "User Presence Polling", "Returns the recommended user presence polling interval"),
+    AH_PREDEF ("_VAL",    "Pin Configuration Value", "Resource Descriptor field"),
```
Comment (zetalog) [inline source/common/ahpredef.c]: No Pin Configuration type _TYP?

File: source/compiler/aslcompiler.l [hunk 13]
```diff
@@ -455,6 +455,7 @@ NamePathTail                [.]{NameSeg}
 "Memory24"                  { count (1); return (PARSEOP_MEMORY24); }
 "Memory32"                  { count (1); return (PARSEOP_MEMORY32); }
 "Memory32Fixed"             { count (1); return (PARSEOP_MEMORY32FIXED); }
+"PinConfig"                 { count (1); return (PARSEOP_PIN_CONFIG); }
```
Comment (zetalog) [inline source/compiler/aslcompiler.l]: Better naming style: PARSEOP_PINCONFIG like other resource descriptors.

**Comment by acpibob** [comment]
This table is used for dispatch and must be in order of the value.

AcpiGbl_DmResourceDispatch [ResourceIndex]

**Comment by acpibob** [comment]
Don’t forget that the legal header has been amended for version 20170303.

**Pull request merged**